crates/apub/src/http/inbox.rs

```diff
@@ -0,0 +1,171 @@
+      // in correct order even when delivered out of order.
+      let request_parts = (
+        request.headers().clone(),
+        request.method().clone(),
+        request.uri().clone(),
+      );
+      ACTIVITY_QUEUE.write().unwrap().push(InboxActivity {
+        request_parts,
+        bytes,
+        published,
+      });
+      Ok(HttpResponse::Ok().finish())
+    }
+    None => {
+      // no timestamp included, process immediately
+      receive_activity::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+        request, bytes, &data,
+      )
+      .await
+    }
+  }
+}
+
+/// Queue of incoming activities, ordered by oldest published first
+static ACTIVITY_QUEUE: Lazy<Arc<RwLock<BinaryHeap<InboxActivity>>>> =
+  Lazy::new(|| Arc::new(RwLock::new(BinaryHeap::new())));
+
+/// Minimum age of an activity before it gets processed. This ensures that an activity which was
+/// delayed still gets processed in correct order.
+const RECEIVE_DELAY: Option<TimeDelta> = TimeDelta::try_seconds(1);
+
+pub fn handle_received_activities(
+  context: &Data<LemmyContext>,
+) -> LemmyResult<Vec<JoinHandle<()>>> {
+  // launch one task per cpu core
+  let parallelism = available_parallelism()?.into();
+  let workers = (0..parallelism)
+    .map(|_| {
+      let context = context.reset_request_count();
+      spawn(async move {
+        loop {
+          let now = Local::now();
+          if let Some(latest_timestamp) = peek_queue_timestamp() {
+            if latest_timestamp < now - RECEIVE_DELAY.unwrap() {
+              if let Some(a) = pop_queue() {
+                let parts = (&a.request_parts.0, &a.request_parts.1, &a.request_parts.2);
+                receive_activity_parts::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+                  parts, a.bytes, &context,
+                )
+                .await
+                .inspect_err(|e| info!("Error receiving activity: {e}"))
+                .ok();
+              }
+            }
+          }
+          // TODO: could sleep based on remaining time until head activity reaches 1s
+          //       or simply use `WORK_FINISHED_RECHECK_DELAY` from lemmy_federate
+          sleep(Duration::from_millis(100)).await;
+          // TODO: need cancel? lemmy seems to shutdown just fine
+        }
+      })
+    })
+    .collect();
+
+  Ok(workers)
+}
+
+fn peek_queue_timestamp() -> Option<DateTime<Utc>> {
+  ACTIVITY_QUEUE.read().unwrap().peek().map(|i| i.published)
+}
+
+fn pop_queue<'a>() -> Option<InboxActivity> {
+  ACTIVITY_QUEUE.write().unwrap().pop()
+}
+
+#[derive(Clone, Debug)]
+struct InboxActivity {
+  // Need to store like this because HttpRequest is not Sync
+  request_parts: (HeaderMap, Method, Uri),
+  bytes: Bytes,
+  published: DateTime<Utc>,
+}
+
+impl PartialOrd for InboxActivity {
+  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+    other.published.partial_cmp(&self.published)
```

Comment:
There is a test case below. The heap normally has largest items first, so I reverted it here to get smallest items first, so that older activities are processed before newer ones.